feat: Editorial review, Relative color syntax (#32004)
mdn/content: files/en-us/web/css/color_value/rgb/index.md

slug: Web/CSS/color_value/rgb
page-type: css-function
browser-compat: css.types.color.rgb
spec-urls:
  - https://drafts.csswg.org/css-color-5/#relative-RGB
  - https://drafts.csswg.org/css-color/#rgb-functions
---

{{CSSRef}}

The **`rgb()`** functional notation expresses a color in the {{glossary("RGB", "sRGB")}} {{glossary("color space")}} according to its red, green, and blue components. An optional alpha component represents the color's transparency.

{{EmbedInteractiveExample("pages/css/function-rgb.html")}}

## Syntax

```css
/* Absolute values */
rgb(255 255 255)
rgb(255 255 255 / 50%)

/* Relative values */
rgb(from green r g b / 0.5)
rgb(from #0000FF calc(r + 40) calc(g + 40) b)
rgb(from hwb(120deg 10% 20%) r g calc(b + 200))
```

> **Note:** The computed values of relative colors expressed using `rgb()` syntax are {{CSSXref("color_value/color", "color(srgb ... )")}} values.

> > **Note:** The absolute values can also be written in a legacy form in which all values are separated with commas. Additionally, the legacy `rgba()` function can be used, which is an alias for `rgb()` that accepts the same parameters. Mixing number and percent value types is not valid in the comma-separated legacy syntax: the `R`, `G`, and `B` values must be either all numbers or all percentages.

Author: updated

Let's start with an origin color of `hsl(0 100% 50%)` (equivalent to `rgb(255 0 0)`). The following function outputs the same color as the origin color — it uses the origin color's `r`, `g`, and `b` channel values as the output channel values:

```css
rgb(from hsl(0 100% 50%) r g b)
```

This function's output color is the sRGB `color()` equivalent of `rgb(255 0 0)`: `color(srgb 1 0 0)`.

The next function uses absolute values for the output color's channel values, outputting a completely different color not based on the origin color:

```css
rgb(from hsl(0 100% 50%) 132 132 224)
```

In the above case, the output color is the sRGB `color()` equivalent of `rgb(132 132 224)`: `color(srgb 0.517647 0.517647 0.878431)`.

The following function creates a relative color based on the origin color:

```css
rgb(from hsl(0 100% 50%) r 80 80)
```

This example:

- Converts the origin color (`hsl(0 100% 50%)`) into an `rgb()` equivalent (`rgb(255 0 0)`).
- Sets the `R` channel value for the output color to the origin color `rgb()` equivalent's `R` channel value — `255`.

> - Sets the output color's `G` and `B` channel values to new values not based on the origin color: `80` and `80`, respectively. The output `rgb()` color is `rgb(255 80 80)`.

Author: updated

In the examples we've seen so far in this section, the alpha channels have not been explicitly specified, for either the origin or output colors. When the output color alpha channel is not specified, it defaults to the same value as the origin color alpha channel. When the origin color alpha channel is not specified (and it is not a relative color), it defaults to `1`. Therefore, the origin and output alpha channel values are `1` for the above examples.

Let's look at some examples that specify origin and output alpha channel values. The first one specifies the output alpha channel value as being the same as the origin alpha channel value, whereas the second one specifies a different output alpha channel value, unrelated to the origin alpha channel value.

```css
rgb(from hsl(0 100% 50% / 0.8) r g b / alpha)
/* Computed output color: color(srgb 1 0 0 / 0.8) */

rgb(from hsl(0 100% 50% / 0.8) r g b / 0.5)
/* Computed output color: color(srgb 1 0 0 / 0.5) */
```

> In the following example, the `hsl()` origin color is again converted into an `rgb()` representation — `rgb(255 0 0)`. {{cssxref("calc")}} calculations are applied to the `R`, `G`, `B`, and `A` values, and the final output color is `color(srgb 0.5 0.0980392 0.686275 / 0.9)`:

Author: Good call, added!

```html hidden
<div id="container">
  <div class="item" id="one"></div>
  <div class="item" id="two"></div>
  <div class="item" id="three"></div>
</div>
```

#### CSS

```css hidden
#container {
  display: flex;
  width: 100vw;
  height: 100vh;
  box-sizing: border-box;
}

.item {
  flex: 1;
  margin: 20px;
}
```

```css
:root {
  --base-color: orange;
}

#one {
  background-color: var(--base-color);
}

#two {
  background-color: rgb(from var(--base-color) calc(r - 76.5) g calc(b + 76.5));
  /* 76.5 is 30% of 255 */
}

#three {
  background-color: rgb(from var(--base-color) calc(r - 153) g calc(b + 153));
  /* 153 is 60% of 255 */
}

/* Use @supports to add in support for old syntax that requires r g b values to
   be specified as percentages (with units) in calculations. This is required
   for Safari 16.4+. */
@supports (color: rgb(from red r g calc(b + 30%))) {
  #two {
    background-color: rgb(from var(--base-color) calc(r - 30%) g calc(b + 30%));
  }

  #three {
    background-color: rgb(from var(--base-color) calc(r - 60%) g calc(b + 60%));
  }
}
```

#### Result

The output is as follows:

{{ EmbedLiveSample("Using relative colors with rgb()", "100%", "200") }}

> ### Legacy syntax: comma-separated values

Author: nice one

## Syntax

```css
/* Absolute values */

rgb(255 255 255)
rgb(255 255 255 / 50%)

/* Relative values */

rgb(from green r g b / 0.5)
rgb(from #0000FF calc(r + 40) calc(g + 40) b)
rgb(from hwb(120deg 10% 20%) r g calc(b + 200))
```

> The absolute syntax can also be written in a legacy form in which all values are separated with commas.

Author: Updated!

### Values

Below are descriptions of the allowed values for both absolute and [relative colors](/en-US/docs/Web/CSS/CSS_colors/Relative_colors).

> **Note:** Absolute and relative functional notation serializes to sRGB values, and the values of the red, green, blue components may be rounded in serialization.

> > **Note:** See [Missing color components](/en-US/docs/Web/CSS/color_value#missing_color_components) for more information on the effect of `none`.

Author: Really good point — updated!

## Syntax

```css
/* Absolute values */

rgb(255 255 255)
rgb(255 255 255 / 50%)

/* Relative values */

rgb(from green r g b / 0.5)
rgb(from #0000FF calc(r + 40) calc(g + 40) b)
rgb(from hwb(120deg 10% 20%) r g calc(b + 200))
```

> The absolute values can also be written in a legacy form in which all values are separated with commas.

Author: Seems fine to me; updated.

> The absolute values can also be written in a legacy form in which all values are separated with commas. Mixing number and percent value types is not valid in the comma-separated legacy syntax: the `R`, `G`, and `B` values must be either all numbers or all percentages.

Author: I like this. I've done what you suggest (including turning this bit into a note), plus I've also added a bit in about the legacy `rgba()` function, which was covered in the note we've now removed. 

I've repeated the same update on the hsl page.

- The `r`, `g` and `b` values are each resolved to `<number>`s between 0 and 255.
- The `alpha` channel is resolved to a `<number>` between 0 and 1.

When defining a relative color, the different channels of the output color can be expressed in several different ways. Below, we'll study some examples to illustrate these.

It is important to bear in mind that, in the first two examples below, we are using relative color syntax. However, the first one outputs the same color as the origin color and the second one outputs a color not based on the origin color at all. They don't really create relative colors! You'd be unlikely to ever use these in a real codebase, and would probably just use an absolute color value instead. We included these examples as a starting point for learning about relative color syntax.

> Let's start with an origin color of `hsl(0 100% 50%)` (equivalent to `red`). The following function outputs the same color as the origin color — it uses the origin color's `r`, `g`, and `b` channel values as the output channel values:

Reviewer: ```suggestion
Let's start with an origin color of `hsl(0 100% 50%)` (equivalent to `red`). The following function outputs the same color as the origin color. It converts the `hsl()` origin color to `rgb()`, then uses the origin's `r`, `g`, and `b` channel values as the output channel values:
```
Does it make sense to update it this way? Then, we can remove half, if not all, of the note after the example line of code.

Author: I think I prefer to keep it as-is, with the note afterward explaining all of this. As you'll see from some further changes I've made to those notes, in the case of `hsl()`/`hwb()`/`rgb()` the color converted to in the background is actually the sRGB `color()` equivalent. It is a bit more complex than I first thought!

```css
rgb(from hsl(0 100% 50%) r g b)
```

> > **Note:** As mentioned above, if the output color is using a different color model to the origin color, the origin color is converted to the same model as the output color in the background so that it can be represented in a way that is compatible (i.e. using the same channels). For example, in the above case the {{cssxref("color_value/hsl", "hsl()")}} color `hsl(0 100% 50%)` is converted to `rgb(255 0 0)`.

Reviewer: ```suggestion
In this case, the {{cssxref("color_value/hsl", "hsl()")}} color `hsl(0 100% 50%)` is converted to `rgb(255 0 0)`.
```
If we go with the previous edit, all that is needed here is the output color.

Author: See above comment.

This function uses absolute values for the output color's channel values, outputting a completely different color not based on the origin color:

```css
rgb(from hsl(0 100% 50%) 132 132 224)
```

> 

Reviewer: ```suggestion
In this case, the output color is `rgb(132 132 224)`.

```

Author: Except that it isn't. It's actually `color(srgb 0.517647 0.517647 0.878431)`. I'll add these lines in on each color function page, but I'm curious as to what you think about the `color()` equivalents and how confusing they are, in the case of hwb/hsl/rgb

Reviewer 2: I can understand how these are confusing.
Maybe this is a specific aspect that needs to be called out?

The underlying issue is that `color(srgb ...)` can express colors outside of the srgb gamut while `rgb`, `hsl`, ... cannot in a safe way. These older color functions have parse time clamping. So `rgb(300 0 0)` is clamped to `rgb(255 0 0)`.

When round tripping values it would be problematic if relative `rgb()` was serialized to `rgb()`

- `rgb(from red calc(r * 2) g b)`
- serialize as `rgb` -> `"rgb(500 0 0)"` (let's assume that 500 is ok here)
- re-assign to a prop -> `foo.color = "rgb(500 0 0)"`
- parse time clamping rules apply -> `rgb(255 0 0)`

compared to :

- `rgb(from red calc(r * 2) g b)`
- serialize as `color` -> `"color(srgb 2 0 0)"`
- re-assign to a prop -> `foo.color = "color(srgb 2 0 0)"`
- no channels are clamped during parsing -> `rgb(500 0 0)`

-------

This is not something most CSS authors will ever have to think about.
It is only relevant for authors that manipulate or test against serialized values.

The following function uses the origin color's `r` channel value for the output color's `r` channel value, but uses a new value for the output color's `g` and `b` channel values, creating a relative color based on the origin color:

```css
rgb(from hsl(0 100% 50%) r 80 80)
```

> 

Reviewer: ```suggestion

 In this case,`hsl(0 100% 50%)` is converted to `rgb(255 0 0)`. The red remains the same (`r`) green and blue values are replaced by `80`, resolving to `rgb(255 80 80)`.
 
```

Author: I've made an equivalent change that solves for this one, equivalent to the change made on the `lab()` page.

The following example uses {{cssxref("calc")}} functions to calculate new channel values for the output color that are relative to the origin color channel values:

```css
rgb(from hsl(0 100% 50%) calc(r/2) calc(g + 25) calc(b + 175) / calc(alpha - 0.1))
```

> 

Reviewer: In this case,`hsl(0 100% 50%)` is converted to `rgb(255 0 0 / 1)`. After calculating new `r`, `g`, `b`, and `alpha` values using `calc()` functions, the output color is `rgb(127.5 25 175 / 0.9)`. Note that if a calculation were to create values outside of the `0` to `255` bounds, the value would be clamped at `0` or `255`.

Reviewer 2: They are not clamped.
If authors want clamping behavior they need to use the `min/max/clamp` math functions.

`rgb(from hsl(0 100% 50%) r calc(g + 300) b)` is equal to `color(srgb 1 1.17647 0)`.

Notice how the output uses the `color(srgb ...)` function and not `rgb(...)` as `color()` can express values beyond `0 - 1`.

Author: Again, similar change made to the `lab()` page to express what is going on here.

These variants are defined using relative colors — the `--base-color` [custom property](/en-US/docs/Web/CSS/--*) is passed into an `rgb()` function, and the output color has its red and blue channels modified to achieve the desired effect via `calc()` functions, while the green channel is left unchanged.

```html hidden
<div id="container">
  <div class="item" id="one"></div>
  <div class="item" id="two"></div>
  <div class="item" id="three"></div>
</div>
```

#### CSS

```css hidden
#container {
  display: flex;
  width: 100vw;
  height: 100vh;
  box-sizing: border-box;
}

.item {
  flex: 1;
  margin: 20px;
}
```

```css
:root {
  --base-color: orange;
}

#one {
  background-color: var(--base-color);
}

#two {
  /* As per the spec, - and + 30% should be specified like this
  background-color: rgb(from var(--base-color) calc(r - 76.5) g calc(b + 76.5));

  /* In Chrome 121+, r, g, and b channel values incorrectly resolve to numbers
     between 0-1 rather than 0-255, hence + and - 30% currently being specified
     like this */

>   background-color: rgb(from var(--base-color) calc(r - 0.3) g calc(b + 0.3));

Reviewer: Remove the comments from this entire section and instead use:

```css
background-color: rgb(from var(--base-color) calc(r * 0.7) g calc(b * 1.3));
```
I think that will lead to the same result and work per spec.
then add a note about how chrome and safari differ from the spec elsewhere (not in the code block, as comments should just be very brief.

Author: I've got a better fix here - Chrome 121+ supports the spec properly, so I can get rid of most of these comments. I'll update the BCD too.

I left the Safari note in, as I thought just one note looked OK, and not messy.

}

#three {
  /* As per the spec, - and + 60% should be specified like this
  background-color: rgb(from var(--base-color) calc(r - 153) g calc(b + 153));

  /* In Chrome 121+, r, g, and b channel values incorrectly resolve to numbers
     between 0-1 rather than 0-255, hence + and - 60% currently being specified
     like this */
  background-color: rgb(from var(--base-color) calc(r - 0.6) g calc(b + 0.6));
}

/* Use @supports to add in support for Safari 16.4+, which supports old
   syntax that requires r g b values to be specified as percentages (with
   units) in calculations */
@supports (color: rgb(from red r g calc(b + 30%))) {
  #two {
    background-color: rgb(from var(--base-color) calc(r - 30%) g calc(b + 30%));
  }

  #three {
    background-color: rgb(from var(--base-color) calc(r - 60%) g calc(b + 60%));
  }
}
```

#### Result

The output is as follows:

{{ EmbedLiveSample("Using relative colors with rgb()", "100%", "200") }}

### Legacy syntax: comma-separated values

> For legacy reasons, the `rgb()` function accepts a form in which all values are separated using commas.

Reviewer: optionally add the comment about how you can't mix numbers and percents if using commas.

Author: We said this earlier on the same page.

Reviewer: I meant you could add it here instead of earlier. My quick comment was NOT clear. sorry.

Author: No worries.
